submissions/dashakim/a-tiny-world2/index.js

```diff
@@ -0,0 +1,67 @@
+class Inhabitant {
+  constructor(name, gender) {
+    this.resident = this.constructor.name;
+    this.name = name;
+    this.gender = gender;
+  }
+  toSring() {
+    return Object.entries(this)
+      .map(([prop, value]) => `${prop}: <b>${value}</b>`)
+      .join("; ");
+  }
```

Comment:
This method will be inherited by all child classes down to grandchildren and even further. And it would work correctly.

However there is a chance to train a skill of methods overloading. This task is optional. Please let me know if you're gonna to complete it.

So the task per se:

You current implementation relies on the fact that we need all Object properties be communicated to a user. In real business scenarios there will be many properties that we may not want to expose. Imagine, each inhabitant has `age` property that we do not want to expose.

This means we need to list props to print explicitly. This also means that each class must be responsible for printing the props it has knowledge of. Consequently thus means we need to re-define (overload) method `toString` in each sub-class. Yet we can benefit from calling super-class' method to avoid code duplication. `super.methodName()` is to rescue.

Thus we make code well prepared to support more complex taxonomy of Inhabitants and to rely on the structure of information specific to each particular class. This way we follow a principle of Encapsulation that is one of the key principles of OOP.

Can you improve your code accordingly?

Again, you're not required to do so. Just let me know if you take the challenge.

Reply:
Sounds great, yes I'll take the challenge. Also I don't know how to implement the method to add friends without passing into props directly.

Reply:
Friends is all different story/challenge.

A couple of hints:

You should store friends rather than their names. This means you can add friends from among inhabitants (objects) that already exist at the moment of their addition. So, it must be a dedicated method to call.

From the other hand when listing friends we need just their names.

Reply:
Unfortunately, hint doesn't help me. I don't know how to start. Would you so kind to help me? There any possible to write method in existing structure?

Reply:
Partial class implementation for brevity:
```
class Alien {
  constructor() {  this.friends = [];  }
  addFriends(friends) {  this.friends = friends;  }
  getFriendsNames() {  return this.friends.map(friend => friend.name).join(", ");  }
  toString() { return "whatever" + this.getFriendsNames(); }
}

const alien = new Alien(), cat = new Cat(), dog = new Dog();
alien.addFriends([cat, dog]);
console.log("Profile: " + alien);
```